src/FsAutoComplete/LspServers/FSharpLspClient.fs

```diff
@@ -143,74 +144,164 @@ type ServerProgressReport(lspClient: FSharpLspClient, ?token: ProgressToken) =
-
-  override x.OnEventWritten eventArgs =
-
-    lock x
-    <| fun () ->
-      try
-        if isDisposed then
-          ()
-        else
-          let message =
-            match eventArgs.EventId with
-            | 5 ->
-              let progressReport = new ServerProgressReport(lspClient)
-              let message = eventArgs.Payload.[0] |> string
-              let fileName = IO.Path.GetFileName message
-
-              if inflightEvents.TryAdd(eventArgs.Task, progressReport) then
-                progressReport.Begin($"Dependent Typecheck {fileName}", message = message)
-                |> Async.Start
-            | 6 ->
-              let message = eventArgs.Payload.[0] |> string
-
-              match inflightEvents.TryRemove(eventArgs.Task) with
-              | true, report ->
-                report.End($"Finished {message}") |> Async.Start
-                tryDispose report
-                ()
-              | false, _ ->
-
-                ()
-            | other -> ()
-
-          message
-      with e ->
-        ()
-
-  member _.DisableEvents(source) = ``base``.DisableEvents(source)
-
-  interface System.IDisposable with
-    member this.Dispose() =
-      lock this
-      <| fun () ->
-        if isNull source then () else this.DisableEvents(source)
-        isDisposed <- true
-        inflightEvents.Values |> Seq.iter (tryDispose)
-        inflightEvents <- null
+  let inflightEvents =
+    ConcurrentDictionary<_, System.Diagnostics.Activity * ServerProgressReport>()
+
+  let isStopped (activity: Activity) =
+#if NET6_0
+    false
+#else
+    activity.IsStopped
+#endif
+    // giving this 1 seconds to report something, otherwise assume it's a dead activity
+    || ((DateTime.UtcNow - activity.StartTimeUtc) > TimeSpan.FromSeconds(1.)
+        && activity.Duration = TimeSpan.Zero)
+
+  let getTagItemSafe key (a: Activity) = a.GetTagItem key |> Option.ofObj
+
+  let getFileName =
+    getTagItemSafe Tracing.SemanticConventions.FCS.fileName
+    >> Option.map string
+    >> Option.map IO.Path.GetFileName
+    >> Option.defaultValue String.Empty
+
+  let getProject =
+    getTagItemSafe Tracing.SemanticConventions.FCS.project
+    >> Option.map string
+    >> Option.map IO.Path.GetFileName
+    >> Option.defaultValue String.Empty
+
+
+  let getUserOpName =
+    getTagItemSafe Tracing.SemanticConventions.FCS.userOpName
+    >> Option.map string
+    >> Option.defaultValue String.Empty
+
+  let mbp =
+    MailboxProcessor.Start(fun inbox ->
+      async {
+        while not isDisposed do
+          for (a, p) in inflightEvents.Values do
+            // We don't get always get a corresponding ActivityStopped for a given Activity event so this looks at anything that hasn't had a duration incresed at all
+            // We don't seem to get a corresponding ActivityStopped on cancelled typechecks.
+            if isStopped a then
+              do! p.End()
+              inflightEvents.TryRemove(a.Id) |> ignore
```

Comment:
There seems to be an issue with how `activityStopped` doesn't get called with regards to cancelled type checks. It's very easy to reproduce by going to a larger file (like `AdaptiveLSPServer`) and holding down a letter key.  (This will try to type check on every keystroke but cancel if more come in). I'm not sure if this is by design for `ActivityListener` to not call stopped or if there's some weird bug between async/Cancellation/Activities. (I'm just speculating, I have no idea where the issue really is).